protocols/autonatv2/src/client/handler/dial_request.rs

```diff
@@ -0,0 +1,342 @@
+                protocol, ..
+            }) => match self.queued_streams.pop_front() {
+                Some(stream_tx) => {
+                    if let Err(_) = stream_tx.send(Ok(protocol)) {
+                        tracing::warn!("Failed to send stream to dead handler");
+                    }
+                }
+                None => {
+                    tracing::warn!("Opened unexpected substream without a pending dial request");
+                }
+            },
+            ConnectionEvent::RemoteProtocolsChange(ProtocolsChange::Added(mut added)) => {
+                if added.any(|p| p.as_ref() == REQUEST_PROTOCOL_NAME) {
+                    self.queued_events.push_back(
+                        ConnectionHandlerEvent::NotifyBehaviour(ToBehaviour::PeerHasServerSupport)
+                    );
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+async fn start_substream_handle(
+    dial_request: DialRequest,
+    substream_recv: oneshot::Receiver<
+        Result<
+            Stream,
+            StreamUpgradeError<<ReadyUpgrade<StreamProtocol> as OutboundUpgradeSend>::Error>,
+        >,
+    >,
+) -> Result<TestEnd, Error> {
+    match substream_recv.await {
+        Ok(Ok(substream)) => handle_substream(dial_request, substream).await,
+        Ok(Err(err)) => Err(Error::from(err)),
+        Err(_) => Err(Error::InternalServer),
+    }
+}
+
+async fn handle_substream(
+    dial_request: DialRequest,
+    mut substream: impl AsyncRead + AsyncWrite + Unpin,
+) -> Result<TestEnd, Error> {
+    Request::Dial(dial_request.clone())
+        .write_into(&mut substream)
+        .await?;
+    let mut suspicious_addr = Vec::new();
+    loop {
+        match Response::read_from(&mut substream).await? {
+            Response::Data(DialDataRequest {
+                addr_idx,
+                num_bytes,
+            }) => {
+                if addr_idx >= dial_request.addrs.len() {
+                    return Err(Error::InvalidReferencedAddress {
+                        index: addr_idx,
+                        max: dial_request.addrs.len(),
+                    });
+                }
+                if num_bytes > DATA_LEN_UPPER_BOUND {
+                    return Err(Error::DataRequestTooLarge {
+                        len: num_bytes,
+                        max: DATA_LEN_UPPER_BOUND,
+                    });
+                }
+                if num_bytes < DATA_LEN_LOWER_BOUND {
+                    return Err(Error::DataRequestTooSmall {
+                        len: num_bytes,
+                        min: DATA_LEN_LOWER_BOUND,
+                    });
+                }
+                match dial_request.addrs.get(addr_idx) {
+                    Some(addr) => {
+                        tracing::trace!("the address {addr} is suspicious to the server, sending {num_bytes} bytes of data");
+                        suspicious_addr.push(addr.clone());
```

Comment:
I find this rather odd, why do we care at the upper layer that the server made us execute the amplification attack prevention? At the end of day, we care which address got tested and what the outcome was. I'd suggest reporting that as an event back to the behaviour.

Essentially, what the behaviour needs to know is:

- nonce
- address
- outcome (reachable, not reachable, unexpected error)

If you want, you can add a boolean flag for whether AAP was executed but I don't really see why we'd need that.

Reply:
You are right. Correct me if I'm wrong, but I removed that.